Added capabilities for copper chests, shelves, and bundles

thomasglasser opened the pull request:
Abstracted ItemAccessItemHandler to ComponentItemAccessItemHandler and added BundleItemHandler Updated ItemInventoryTests for BundleItemHandler

thomasglasser committed:
Added capabilities for copper chests, shelves, and bundles
Abstracted ItemAccessItemHandler to ComponentItemAccessItemHandler and added BundleItemHandler
Updated ItemInventoryTests for BundleItemHandler

thomasglasser committed:
Merge branch 'neoforged:1.21.x' into 1.21.x-updated-caps

thomasglasser committed:
Fixed ItemInventoryTests

Technici4n commented on src/main/java/net/neoforged/neoforge/capabilities/CapabilityHooks.java:
```diff
@@ -104,7 +115,8 @@ public static void registerVanillaProviders(RegisterCapabilitiesEvent event) {
                 BlockEntityType.CRAFTER,
-                BlockEntityType.DECORATED_POT);
+                BlockEntityType.DECORATED_POT,
+                BlockEntityType.SHELF);
```
An additional patch is required to avoid calling `setChanged` from inside a transaction, i.e. when `setItem` is called on the BE with `insideTransaction` set to `true`.

Technici4n commented on src/main/java/net/neoforged/neoforge/transfer/item/BundleItemHandler.java:
```diff
@@ -0,0 +1,35 @@
+/**
+ * A {@link ComponentItemAccessItemHandler} for {@link BundleContents}.
+ */
+public class BundleItemHandler extends ComponentItemAccessItemHandler<BundleContents> {
```
I don't think that this correctly checks that the items can fit in the bundle before being inserted. It would also allow insertion into arbitrary slots.

Technici4n commented on src/main/java/net/neoforged/neoforge/transfer/item/ComponentItemAccessItemHandler.java:
```diff
@@ -0,0 +1,91 @@
+ * To use this class, register a new {@link DataComponentType} which holds an {@link T} for your item.
+ * Then reference that component from your {@link ICapabilityProvider} passed to {@link RegisterCapabilitiesEvent#registerItem} to create an instance of this class.
+ */
+public abstract class ComponentItemAccessItemHandler<T> extends ItemAccessResourceHandler<ItemResource> {
```
I don't think that this superclass really makes sense, as adding more steps in the hierarchy tends to make things more complicated to understand. Given the changes that will be required for the special behavior of the bundle, better make `ItemAccessItemHandler` and `BundleItemHandler` two subclasses of `ItemAccessResourceHandler`.

Technici4n commented on tests/src/main/java/net/neoforged/neoforge/debug/capabilities/ItemInventoryTests.java:
```diff
@@ -36,11 +41,13 @@ public class ItemInventoryTests {
 
     private static final DeferredItems ITEMS = HELPER.items();
     private static final DeferredItem<Item> BACKPACK;
+    private static final DeferredItem<Item> BUNDLEPACK;
```
Can you use `Items.BUNDLE` instead of registering a new item?

thomasglasser replied:
You previously register a new item instead of using a shulker box so I assumed u wanted to explicitly test custom items.

Technici4n replied:
That is likely carried from back when shulker boxes used `BlockEntityData` which was not compatible with `ComponentItemHandler`.

Technici4n commented on tests/src/main/java/net/neoforged/neoforge/debug/capabilities/ItemInventoryTests.java:
```diff
@@ -58,40 +68,43 @@ static void init(final TestFramework framework) {
 
-            ItemStack storedStick = items.getStackInSlot(STICK_SLOT);
-            helper.assertValueEqual(storedStick.getItem(), Items.STICK, "Default contents should contain a stick at slot " + STICK_SLOT);
+    private static <T> void testItemInventory(ItemStack stack, ExtendedGameTestHelper helper, DataComponentType<T> component, BiFunction<T, Integer, ItemStack> stackGetter) {
```
I am not sure that this makes much sense. For the bundle test you should really test various special cases, such as evolution of the slot count as items are inserted, weight of different items, items that cannot be nested, etc...

Technici4n commented on src/main/resources/META-INF/accesstransformer.cfg:
```diff
@@ -276,6 +276,7 @@ private-f net.minecraft.world.level.storage.loot.LootPool bonusRolls # bonusRoll
 public net.minecraft.server.network.ServerConfigurationPacketListenerImpl finishCurrentTask(Lnet/minecraft/server/network/ConfigurationTask$Type;)V
 public-f net.minecraft.world.entity.npc.VillagerTrades WANDERING_TRADER_TRADES # WANDERING_TRADER_TRADES
 public net.minecraft.world.entity.EquipmentSlot countLimit
+public net.minecraft.world.item.component.ItemContainerContents MAX_SIZE # MAX_SIZE
```
I'd rather not add this AT and instead leave the `/* MAX_SIZE */` comment.

Technici4n reviewed (changes requested):
Sorry for taking so long to review, it completely slipped my mind...

thomasglasser committed:
Merge pull request #1 from neoforged/1.21.x

merge upstream

thomasglasser committed:
Refactored BundleItemHandler and ItemInventoryTests
Reverted ItemAccessItemHandler and accesstransformer.cfg
Added ShelfBlockEntity.java.patch

Technici4n commented on src/main/java/net/neoforged/neoforge/transfer/item/BundleItemHandler.java:
```diff
@@ -0,0 +1,133 @@
+    public int size() {
+        BundleContents contents = itemAccess.getResource().get(component);
+        if (contents == null) return 0;
+        return contents.size() + 1;
```
Do we want to return `contents.size()` if the weight is 1?

Technici4n commented on src/main/java/net/neoforged/neoforge/transfer/item/BundleItemHandler.java:
```diff
@@ -0,0 +1,133 @@
+    @Override
+    public long getAmountAsLong(int index) {
+        if (itemAccess.getResource().is(validItem)) {
+            return getStack(index).getCount();
```
```suggestion
            return (long) itemAccess.getAmount() * getStack(index).getCount();
```

Technici4n commented on src/main/java/net/neoforged/neoforge/transfer/item/BundleItemHandler.java:
```diff
@@ -0,0 +1,133 @@
+
+    @Override
+    public long getCapacityAsLong(int index, ItemResource resource) {
+        return resource.isEmpty() ? Item.ABSOLUTE_MAX_STACK_SIZE : Math.min(resource.getMaxStackSize(), Item.ABSOLUTE_MAX_STACK_SIZE);
```
This should also call `isValid` and multiply by the access amount.

Technici4n commented on src/main/java/net/neoforged/neoforge/transfer/item/BundleItemHandler.java:
```diff
@@ -0,0 +1,133 @@
+    @Override
+    public boolean isValid(int index, ItemResource resource) {
+        // Any resource is valid, but we have to check that the item of the item access has not changed.
+        return itemAccess.getResource().is(validItem);
```
This should check `BundleContents.canItemBeInBundle` as well.

Technici4n commented on src/main/java/net/neoforged/neoforge/transfer/item/BundleItemHandler.java:
```diff
@@ -0,0 +1,133 @@
+    @Override
+    public int insert(int index, ItemResource resource, int amount, TransactionContext transaction) {
+        TransferPreconditions.checkNonEmptyNonNegative(resource, amount);
+        if (index < 0) return 0;
```
You can throw an error if that happens, e.g. with `checkNonNegative`.

Technici4n commented on src/main/java/net/neoforged/neoforge/transfer/item/BundleItemHandler.java:
```diff
@@ -0,0 +1,133 @@
+        ItemResource accessResource = itemAccess.getResource();
+        BundleContents contents = accessResource.get(component);
+        if (contents == null || index > contents.size()) return 0;
+
```
This is missing amount handling, see `ItemAccessResourceHandler` for inspiration. Generally speaking the incoming `amount` needs to be divided by the current amount of the item access.

Technici4n commented on src/main/java/net/neoforged/neoforge/transfer/item/BundleItemHandler.java:
```diff
@@ -0,0 +1,133 @@
+        int inserted = 0;
+
+        if (index == contents.size()) {
+            inserted = mutable.tryInsert(resource.toStack(amount));
```
We should potentially refuse the insertion if the bundle contains the item already? I don't think it's necessary, but we could add at least a comment that such a check could be added in the future.

Technici4n replied:
@thomasglasser remember to update this function for proper amount handling as well

Technici4n commented on src/main/java/net/neoforged/neoforge/transfer/item/BundleItemHandler.java:
```diff
@@ -0,0 +1,133 @@
+        } else {
+            ItemStack existing = contents.getItemUnsafe(index);
+            if (resource.matches(existing)) {
+                inserted += mutable.tryInsert(resource.toStack(Math.min(amount, existing.getMaxStackSize() - existing.getCount())));
```
Is there any reason to query `getMaxStackSize()` here versus just trying to insert `amount`?

Technici4n commented on src/main/java/net/neoforged/neoforge/transfer/item/BundleItemHandler.java:
```diff
@@ -0,0 +1,133 @@
+    public int extract(int index, ItemResource resource, int amount, TransactionContext transaction) {
+        TransferPreconditions.checkNonEmptyNonNegative(resource, amount);
+        if (index < 0) return 0;
+
```
amount should be divided by the current amount in the access.

Technici4n commented on tests/src/main/java/net/neoforged/neoforge/debug/capabilities/ItemInventoryTests.java:
```diff
@@ -24,61 +24,55 @@
+            }, Items.SHULKER_BOX);
+            e.registerItem(Capabilities.Item.ITEM, (stack, itemAccess) -> {
+                return new BundleItemHandler(itemAccess, DataComponents.BUNDLE_CONTENTS);
+            }, Items.BUNDLE);
```
This should be removable since `CapabilityHooks` already handles it

Technici4n commented on tests/src/main/java/net/neoforged/neoforge/debug/capabilities/ItemInventoryTests.java:
```diff
@@ -24,61 +24,55 @@
             e.registerItem(Capabilities.Item.ITEM, (stack, itemAccess) -> {
                 return new ItemAccessItemHandler(itemAccess, DataComponents.CONTAINER, SLOTS);
-            }, BACKPACK);
+            }, Items.SHULKER_BOX);
```
Either keep using the `BACKPACK`, or don't register anything. This seems to override the shulker box's item handler which is not good.

Technici4n commented on src/main/java/net/neoforged/neoforge/transfer/item/BundleItemHandler.java:
```diff
@@ -0,0 +1,133 @@
+ * The stacks are stored in a {@link BundleContents} data component.
+ * <p>
+ * To use this class, register a new {@link DataComponentType} which holds an {@link BundleContents} for your item.
+ * Then reference that component from your {@link ICapabilityProvider} passed to {@link RegisterCapabilitiesEvent#registerItem} to create an instance of this class.
```
Nit: not really a base implementation for everyone to use. This is specific to bundles. So `Item handler implementation for bundles.` or something like that would be a little more accurate.

Technici4n reviewed (changes requested):
A few oversights in the bundle item handler implementation, very good overall!

thomasglasser committed:
Some requested fixes
Re-added ItemInventoryTests backpack to test custom container size

thomasglasser committed:
Merge branch '1.21.x' into 1.21.x-updated-caps

thomasglasser committed:
Ported ShelfBlockEntity.java.patch

thomasglasser committed:
Merge branch '1.21.x' into 1.21.x-updated-caps

thomasglasser committed:
Fixed BundleItemHandler issues

thomasglasser committed:
Merge remote-tracking branch 'origin/1.21.x-updated-caps' into 1.21.x-updated-caps

thomasglasser committed:
Fixed BundleItemHandler#insert

thomasglasser committed:
Merge branch '1.21.x' into 1.21.x-updated-caps

thomasglasser committed:
Added requested TODO

thomasglasser committed:
Merge remote-tracking branch 'origin/1.21.x-updated-caps' into 1.21.x-updated-caps

thomasglasser committed:
Merge branch '1.21.x' into 1.21.x-updated-caps

Technici4n committed:
Use props lambda like before

Technici4n reviewed (approved):
Looks great, many thanks and sorry for the delay!
